File: programs/benchmarks-alioth/python/fannkuchredux.py
# The Computer Language Benchmarks Game
# http://benchmarksgame.alioth.debian.org/
#
# contributed by Miroslav Rubanets
# adapted for  single-threaded p2w
#
# Simple fannkuch-redux implementation


def fannkuch(n: int) -> None:
    # Initialize permutation
    perm: list[int] = []
    i: i32 = 0
    while i < n:
        perm.append(i)
        i = i + 1

    # Initialize count
    count: list[int] = []
    i = 0
    while i < n:
        count.append(0)
        i = i + 1

    max_flips: int = 0
    checksum: int = 0
    sign: int = 1

    perm_count: int = 0
    m: int = n - 1

    # Copy for permutation work
    perm1: list[int] = []
    i = 0
    while i < n:
        perm1.append(i)
        i = i + 1

    while True:
        # Copy perm1 to perm
        i = 0
        while i < n:
            perm[i] = perm1[i]
            i = i + 1

        # Count flips
        k: int = perm[0]
        if k != 0:
            flips: int = 0
            while k != 0:
                # Reverse perm[0:k+1]
                lo: i32 = 0
                hi: int = k
                while lo < hi:
                    tmp: int = perm[lo]
                    perm[lo] = perm[hi]
                    perm[hi] = tmp
                    lo = lo + 1
                    hi = hi - 1
                flips = flips + 1
                k = perm[0]

            if flips > max_flips:
                max_flips = flips
            checksum = checksum + sign * flips

        sign = 0 - sign

        # Generate next permutation
        # Rotate perm1
        tmp: int = perm1[1]
        perm1[1] = perm1[0]
        perm1[0] = tmp

        i = 1
        count[i] = count[i] + 1
        while count[i] > i:
            count[i] = 0
            i = i + 1
            if i >= n:
                print(checksum)
                print("Pfannkuchen(" + str(n) + ") =", max_flips)
                return

            # Rotate perm1[0:i+1]
            tmp = perm1[0]
            j: i32 = 0
            while j < i:
                perm1[j] = perm1[j + 1]
                j = j + 1
            perm1[i] = tmp

            count[i] = count[i] + 1


def main(n: int) -> None:
    fannkuch(n)


# Default argument for testing
main(7)

File: programs/benchmarks-alioth/zig/fannkuchredux.zig
// The Computer Language Benchmarks Game
// http://benchmarksgame.alioth.debian.org/
//
// Zig port (Zig 0.15+)

const std = @import("std");
const posix = std.posix;

fn print(comptime fmt: []const u8, args: anytype) void {
    var buf: [256]u8 = undefined;
    const msg = std.fmt.bufPrint(&buf, fmt, args) catch unreachable;
    _ = posix.write(posix.STDOUT_FILENO, msg) catch {};
}

fn fannkuchredux(n: usize) !struct { checksum: i32, max_flips: i32 } {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const perm = try allocator.alloc(i32, n);
    defer allocator.free(perm);
    const perm1 = try allocator.alloc(i32, n);
    defer allocator.free(perm1);
    const count = try allocator.alloc(i32, n);
    defer allocator.free(count);

    var max_flips_count: i32 = 0;
    var perm_count: i32 = 0;
    var checksum: i32 = 0;

    for (0..n) |i| {
        perm1[i] = @intCast(i);
    }

    var r: usize = n;

    outer: while (true) {
        while (r != 1) {
            count[r - 1] = @intCast(r);
            r -= 1;
        }

        for (0..n) |i| {
            perm[i] = perm1[i];
        }

        var flips_count: i32 = 0;

        while (true) {
            const k: usize = @intCast(perm[0]);
            if (k == 0) break;

            const k2 = (k + 1) >> 1;
            for (0..k2) |i| {
                const temp = perm[i];
                perm[i] = perm[k - i];
                perm[k - i] = temp;
            }
            flips_count += 1;
        }

        max_flips_count = @max(max_flips_count, flips_count);
        if (@mod(perm_count, 2) == 0) {
            checksum += flips_count;
        } else {
            checksum -= flips_count;
        }

        while (true) {
            if (r == n) {
                return .{ .checksum = checksum, .max_flips = max_flips_count };
            }

            const perm0 = perm1[0];
            var i: usize = 0;
            while (i < r) {
                const j = i + 1;
                perm1[i] = perm1[j];
                i = j;
            }
            perm1[r] = perm0;
            count[r] -= 1;
            if (count[r] > 0) break;
            r += 1;
        }
        perm_count += 1;
        continue :outer;
    }
}

pub fn main() !void {
    var args = std.process.args();
    _ = args.skip();

    const n_str = args.next() orelse "7";
    const n = std.fmt.parseInt(usize, n_str, 10) catch 7;

    const result = try fannkuchredux(n);

    print("{d}\n", .{result.checksum});
    print("Pfannkuchen({d}) = {d}\n", .{ n, result.max_flips });
}
